docs_src/design/adr/core/0021-Device-Profile-Changes.md

```diff
@@ -0,0 +1,64 @@
+- Device resources or device commands can be added to a device profile anytime
+- Attributes and properties can be added to a device profile anytime
+- A device profile can be modified when the device profile is not associated to a device or provision watcher
+    - this includes modifying any properties (except identifiers like names and ids)
+    - this includes changes to the array of device resources, device commands
+    - this includes changes to attributes and properties (of device resources)
+    - even when a device profile is associated to a device or provision watcher, properties of the device profile or device resource can be modified when the property change will not affect the behavior of the system (example: the description property has no effect on the system).
+- A device profile cannot be removed when it is associated to a device or provision watcher.
+- A device profile can be removed or modified even when associated to an event or reading unless configuration options (see New Configuration Settings below) are set prohibiting removal or modification when associated to an event/reading
+    - the rationale behind this is that events and readings are generally considered short lived (ephemeral) objects and already contain the necessary device profile information that are needed by the system during their short life without having to refer to the device profile
+    - see note below in Consequences that a new Units property must be added to the Reading object in order to support this rule and the need for all relevant profile data to be in the event/reading.
+
+### Ancillary Rules associated to Device Profiles
+- Identifying or “key” fields for device profiles, device resources, etc. cannot be modified and can never be null.
+- A device profile can begin life “empty” - meaning that it has no device resources or device commands.
+
+### New APIs
+
+The following APIs would be added to the metadata REST service in order to meet the design specified above.
+
+- Add Profile General Property PATCH API (allow to modify profile properties except name and id)
+- Add Profile Device Resource POST API
+- Add Profile Device Resource PATCH API (allow to modify Description and IsHidden only)
+- Add Profile Device Resource DELETE API (allow as described above)
+- Add Profile Device Command POST API
+- Add Profile Device Command PATCH API (allow as described above)
+- Add Profile Device Command DELETE API (allow as described above)
+
+### New Configuration Settings
+
+Some adopters may not view event/reading data as ephemeral or short lived.  These adopters may choose not to allow device profiles to be modified or removed when associated to an event or reading.
+For this reason, two new configuration options will be added to metadata configuration that are used to reject modifications or deletions.
+
+- AllowEventAssociatedDeviceProfileModifications (set to true by default)
+- AllowEventAssociatedDeviceProfileDeletes (set to true by default)
```

Comment:
might be better to rename to `AllowEventAssociatedDeviceProfileDeletions` (noun)

Reply:
The new names are super long, how about `StrictDeviceProfileChanges` and `StringDeviceProfileDeletes`.

Reply:
Do we want these in [Writable] so they can be changed dynamically incase the user needs to temporary switch to make a change?

Reply:
yeah, shorter name is better
it should be `StrictDeviceProfileDeletes`

Reply:
made shorter and put in Writable (proposed Writable.ProfileChange section).